Import tests fail due to unknown provider with Terraform v0.12-alpha2

<!--
Hi there,

Thank you for opening an issue. Please note that we try to keep the Terraform issue tracker reserved for bug reports and feature requests. For general usage questions, please see: https://www.terraform.io/community.html.

If your issue relates to a specific Terraform provider, please open it in the provider's own repository. The index of providers is at https://github.com/terraform-providers .
-->

### Terraform Version
<!---
Run `terraform -v` to show the version, and paste the result between the ``` marks below.

If you are not running the latest version of Terraform, please try upgrading because your issue may have already been fixed.
-->

```
2018/11/20 11:56:45 [INFO] Terraform version: 0.12.0 alpha2
2018/11/20 11:56:45 [INFO] Go runtime version: go1.11.1
```

### Terraform Configuration Files
<!--
Paste the relevant parts of your Terraform configuration between the ``` marks below.

For large Terraform configs, please use a service like Dropbox and share a link to the ZIP file. For security, you can also encrypt the files using our GPG public key.
-->

```hcl
	# Need to have this block even though it's empty; for import testing
	provider "oci" {
	}
	variable "tenancy_ocid" {
		default = "` + getEnvSettingWithBlankDefault("tenancy_ocid") + `"
	}

	variable "ssh_public_key" {
		default = "ssh-rsa"
	}

	variable "region" {
		default = "` + getEnvSettingWithBlankDefault("region") + `"
	}

import id \"ocid1.saml2idp.oc1..aaaaaaaada6m4vhbwpg3wymmoqrlcttne2ryqw7qudwd2e56chpdao4mwh4a\"
```

### Debug Output
<!--
Full debug output can be obtained by running Terraform with the environment variable `TF_LOG=trace`. Please create a GitHub Gist containing the debug output. Please do _not_ paste the debug output in the issue, since debug output is long.

Debug output may contain sensitive information. Please review it before posting publicly, and if you are concerned feel free to encrypt the files using the HashiCorp security public key.
-->
2018/11/20 10:33:34 [DEBUG] Test: Executing step 5
2018/11/20 10:33:34 [DEBUG] provider has no plugin.Client
2018/11/20 10:33:34 [DEBUG] adding implicit provider configuration `provider.<invalid>`, implied first by oci_identity_identity_provider.test_identity_provider (import id "ocid1.saml2idp.oc1..aaaaaaaada6m4vhbwpg3wymmoqrlcttne2ryqw7qudwd2e56chpdao4mwh4a")
2018/11/20 10:33:34 [DEBUG] ProviderTransformer: "oci_identity_identity_provider.test_identity_provider (import id \"ocid1.saml2idp.oc1..aaaaaaaada6m4vhbwpg3wymmoqrlcttne2ryqw7qudwd2e56chpdao4mwh4a\")" (*terraform.graphNodeImportState) needs provider.<invalid>
2018/11/20 10:33:34 [DEBUG] pruning unused provider.oci
2018/11/20 10:33:34 [ERROR] AttachSchemaTransformer: No provider config schema available for provider.<invalid>
2018/11/20 10:33:34 [DEBUG] ReferenceTransformer: "var.region" references: []
2018/11/20 10:33:34 [DEBUG] ReferenceTransformer: "var.tenancy_ocid" references: []
2018/11/20 10:33:34 [DEBUG] ReferenceTransformer: "provider.<invalid>" references: []
2018/11/20 10:33:34 [DEBUG] ReferenceTransformer: "oci_identity_identity_provider.test_identity_provider (import id \"ocid1.saml2idp.oc1..aaaaaaaada6m4vhbwpg3wymmoqrlcttne2ryqw7qudwd2e56chpdao4mwh4a\")" references: []
2018/11/20 10:33:34 [DEBUG] ReferenceTransformer: "var.ssh_public_key" references: []
2018/11/20 10:33:34 [DEBUG] Starting graph walk: walkImport
2018/11/20 10:33:34 [ERROR] <root>: eval: *terraform.EvalInitProvider, err: unknown provider ""
2018/11/20 10:33:34 [ERROR] <root>: eval: *terraform.EvalSequence, err: unknown provider ""
2018/11/20 10:33:34 [ERROR] Test: ImportState failure: unknown provider ""
2018/11/20 10:33:34 [WARN] Test: Executing destroy step

### Crash Output
<!--
If the console output indicates that Terraform crashed, please share a link to a GitHub Gist containing the output of the `crash.log` file.
-->

### Expected Behavior
<!--
What should have happened?
-->
Import tests should succeed (in line with behavior at 0.11.x)
### Actual Behavior
<!--
What actually happened?
-->
The debug output shows that the provider was treated as invalid, `provider.<invalid>` on the importing resource., this further led to pruning of the unused `provider.oci` and later fail with `unknown provider ""`

### Steps to Reproduce
<!--
Please list the full steps required to reproduce the issue, for example:
1. `terraform init`
2. `terraform apply`
-->
TF_ACC=1 TF_LOG=DEBUG OCI_GO_SDK_DEBUG=1 go test ./oci -v   -run TestIdentityIdentityProviderResource_basic -timeout  120m

### Additional Context
<!--
Are there anything atypical about your situation that we should know? For example: is Terraform running in a wrapper script or in a CI system? Are you passing any unusual command line options or environment variables to opt-in to non-default behavior?
-->
This particular test: https://github.com/terraform-providers/terraform-provider-oci/blob/master/oci/identity_identity_provider_test.go#L190 has 6 test steps: Create, Delete, CreateWithOptionals, Update, DataSource and Import.

The 6th test step for import is what fails with this error likely due to pruning of unused provider.

### References
<!--
Are there any other GitHub issues (open or closed) or Pull Requests that should be linked here? For example:

- #6017

-->

--- a/helper/resource/testing_config.go
+++ b/helper/resource/testing_config.go
@@ -45,7 +45,11 @@ func testStep(opts terraform.ContextOpts, state *terraform.State, step TestStep,
 
 	// Build the context
 	opts.Config = cfg
-	opts.State = terraform.MustShimLegacyState(state)
+	opts.State, err = terraform.ShimLegacyState(state)
+	if err != nil {
+		return nil, err
+	}
+
 	opts.Destroy = step.Destroy
 	ctx, stepDiags := terraform.NewContext(&opts)
 	if stepDiags.HasErrors() {
@@ -61,7 +65,11 @@ func testStep(opts terraform.ContextOpts, state *terraform.State, step TestStep,
 
 	// Refresh!
 	newState, stepDiags := ctx.Refresh()
-	state = mustShimNewState(newState, schemas)
+	// shim the state first so the test can check the state on errors
+	state, err = shimNewState(newState, schemas)
+	if err != nil {
+		return nil, err
+	}
 	if stepDiags.HasErrors() {
 		return state, fmt.Errorf("Error refreshing: %s", stepDiags.Err())
 	}
@@ -83,7 +91,11 @@ func testStep(opts terraform.ContextOpts, state *terraform.State, step TestStep,
 
 		// Apply the diff, creating real resources.
 		newState, stepDiags = ctx.Apply()
-		state = mustShimNewState(newState, schemas)
+		// shim the state first so the test can check the state on errors
+		state, err = shimNewState(newState, schemas)
+		if err != nil {
+			return nil, err
+		}
 		if stepDiags.HasErrors() {
 			return state, fmt.Errorf("Error applying: %s", stepDiags.Err())
 		}
@@ -123,7 +135,11 @@ func testStep(opts terraform.ContextOpts, state *terraform.State, step TestStep,
 		if stepDiags.HasErrors() {
 			return state, fmt.Errorf("Error on follow-up refresh: %s", stepDiags.Err())
 		}
-		state = mustShimNewState(newState, schemas)
+
+		state, err = shimNewState(newState, schemas)
+		if err != nil {
+			return nil, err
+		}
 	}
 	if p, stepDiags = ctx.Plan(); stepDiags.HasErrors() {
 		return state, fmt.Errorf("Error on second follow-up plan: %s", stepDiags.Err())

--- a/helper/resource/testing_import_state.go
+++ b/helper/resource/testing_import_state.go
@@ -91,7 +91,10 @@ func testStepImportState(
 		return state, stepDiags.Err()
 	}
 
-	newState := mustShimNewState(importedState, schemas)
+	newState, err := shimNewState(importedState, schemas)
+	if err != nil {
+		return nil, err
+	}
 
 	// Go through the new state and verify
 	if step.ImportStateCheck != nil {
@@ -127,13 +130,31 @@ func testStepImportState(
 					r.Primary.ID)
 			}
 
+			// don't add empty flatmapped containers, so we can more easily
+			// compare the attributes
+			skipEmpty := func(k, v string) bool {
+				if strings.HasSuffix(k, ".#") || strings.HasSuffix(k, ".%") {
+					if v == "0" {
+						return true
+					}
+				}
+				return false
+			}
+
 			// Compare their attributes
 			actual := make(map[string]string)
 			for k, v := range r.Primary.Attributes {
+				if skipEmpty(k, v) {
+					continue
+				}
 				actual[k] = v
 			}
+
 			expected := make(map[string]string)
 			for k, v := range oldR.Primary.Attributes {
+				if skipEmpty(k, v) {
+					continue
+				}
 				expected[k] = v
 			}
 

--- a/terraform/context.go
+++ b/terraform/context.go
@@ -901,14 +901,14 @@ func parseVariableAsHCL(name string, input string, targetType config.VariableTyp
 	}
 }
 
-// shimLegacyState is a helper that takes the legacy state type and
+// ShimLegacyState is a helper that takes the legacy state type and
 // converts it to the new state type.
 //
 // This is implemented as a state file upgrade, so it will not preserve
 // parts of the state structure that are not included in a serialized state,
 // such as the resolved results of any local values, outputs in non-root
 // modules, etc.
-func shimLegacyState(legacy *State) (*states.State, error) {
+func ShimLegacyState(legacy *State) (*states.State, error) {
 	if legacy == nil {
 		return nil, nil
 	}
@@ -928,7 +928,7 @@ func shimLegacyState(legacy *State) (*states.State, error) {
 // the conversion does not succeed. This is primarily intended for tests where
 // the given legacy state is an object constructed within the test.
 func MustShimLegacyState(legacy *State) *states.State {
-	ret, err := shimLegacyState(legacy)
+	ret, err := ShimLegacyState(legacy)
 	if err != nil {
 		panic(err)
 	}

--- a/terraform/context_import_test.go
+++ b/terraform/context_import_test.go
@@ -37,8 +37,7 @@ func TestContextImport_basic(t *testing.T) {
 				Addr: addrs.RootModuleInstance.ResourceInstance(
 					addrs.ManagedResourceMode, "aws_instance", "foo", addrs.NoKey,
 				),
-				ID:           "bar",
-				ProviderAddr: addrs.RootModuleInstance.ProviderConfigDefault("aws"),
+				ID: "bar",
 			},
 		},
 	})
@@ -77,8 +76,7 @@ func TestContextImport_countIndex(t *testing.T) {
 				Addr: addrs.RootModuleInstance.ResourceInstance(
 					addrs.ManagedResourceMode, "aws_instance", "foo", addrs.IntKey(0),
 				),
-				ID:           "bar",
-				ProviderAddr: addrs.RootModuleInstance.ProviderConfigDefault("aws"),
+				ID: "bar",
 			},
 		},
 	})
@@ -135,8 +133,7 @@ func TestContextImport_collision(t *testing.T) {
 				Addr: addrs.RootModuleInstance.ResourceInstance(
 					addrs.ManagedResourceMode, "aws_instance", "foo", addrs.NoKey,
 				),
-				ID:           "bar",
-				ProviderAddr: addrs.RootModuleInstance.ProviderConfigDefault("aws"),
+				ID: "bar",
 			},
 		},
 	})
@@ -179,8 +176,7 @@ func TestContextImport_missingType(t *testing.T) {
 				Addr: addrs.RootModuleInstance.ResourceInstance(
 					addrs.ManagedResourceMode, "aws_instance", "foo", addrs.NoKey,
 				),
-				ID:           "bar",
-				ProviderAddr: addrs.RootModuleInstance.ProviderConfigDefault("aws"),
+				ID: "bar",
 			},
 		},
 	})
@@ -233,8 +229,7 @@ func TestContextImport_moduleProvider(t *testing.T) {
 				Addr: addrs.RootModuleInstance.ResourceInstance(
 					addrs.ManagedResourceMode, "aws_instance", "foo", addrs.NoKey,
 				),
-				ID:           "bar",
-				ProviderAddr: addrs.RootModuleInstance.ProviderConfigDefault("aws"),
+				ID: "bar",
 			},
 		},
 	})
@@ -291,8 +286,7 @@ func TestContextImport_providerModule(t *testing.T) {
 				Addr: addrs.RootModuleInstance.Child("child", addrs.NoKey).ResourceInstance(
 					addrs.ManagedResourceMode, "aws_instance", "foo", addrs.NoKey,
 				),
-				ID:           "bar",
-				ProviderAddr: addrs.RootModuleInstance.ProviderConfigDefault("aws"),
+				ID: "bar",
 			},
 		},
 	})
@@ -349,8 +343,7 @@ func TestContextImport_providerVarConfig(t *testing.T) {
 				Addr: addrs.RootModuleInstance.ResourceInstance(
 					addrs.ManagedResourceMode, "aws_instance", "foo", addrs.NoKey,
 				),
-				ID:           "bar",
-				ProviderAddr: addrs.RootModuleInstance.ProviderConfigDefault("aws"),
+				ID: "bar",
 			},
 		},
 	})

--- a/terraform/test-fixtures/import-provider-resource/main.tf
+++ b/terraform/test-fixtures/import-provider-resource/main.tf
@@ -1,0 +1,7 @@
+provider "aws" {
+  foo = data.template_data_source.d.foo
+}
+
+data "template_data_source" "d" {
+  foo = "bar"
+}

--- a/terraform/transform_import_state.go
+++ b/terraform/transform_import_state.go
@@ -16,10 +16,18 @@ type ImportStateTransformer struct {
 
 func (t *ImportStateTransformer) Transform(g *Graph) error {
 	for _, target := range t.Targets {
+		// The ProviderAddr may not be supplied for non-aliased providers.
+		// This will be populated if the targets come from the cli, but tests
+		// may not specify implied provider addresses.
+		providerAddr := target.ProviderAddr
+		if providerAddr.ProviderConfig.Type == "" {
+			providerAddr = target.Addr.Resource.Resource.DefaultProviderConfig().Absolute(target.Addr.Module)
+		}
+
 		node := &graphNodeImportState{
 			Addr:         target.Addr,
 			ID:           target.ID,
-			ProviderAddr: target.ProviderAddr,
+			ProviderAddr: providerAddr,
 		}
 		g.Add(node)
 	}

